File: backend/src/ssm/ssh/script_runner.py
"""Drive ``script.sh`` on the remote host for every authorized_keys op.

The script is the Rust backend's shell probe (bundled here verbatim). It does
three things we cannot express in plain SFTP:

1. **Home-directory lookup** — via ``getent passwd`` with an ``/etc/passwd``
   fallback — so logins on BSD / TrueNAS / pfSense / anything non-Linux still
   find their ``~/.ssh/authorized_keys``.
2. **Readonly probe + vendor fingerprints** — ``system_readonly``,
   ``user_readonly``, pfSense, TrueNAS Core/Scale, Sophos UTM.
3. **``has_pragma`` detection** — tells the UI whether a file was last written
   by ssm vs. edited by hand.

All writes to ``authorized_keys`` MUST go through
:meth:`ScriptRunner.set_authorized_keyfile` — never ``write_file`` directly —
so the script can back up handwritten files and enforce the readonly flag.
"""

from __future__ import annotations

import hashlib
import json
import shlex
from dataclasses import dataclass
from importlib import resources
from typing import Any

from ssm.core.errors import SshConnectFailed, SshReadOnly
from ssm.ssh.protocol import SshClient, SshTarget

REMOTE_SCRIPT_PATH = ".ssm/script.sh"
_SCRIPT_PACKAGE = "ssm.ssh"
_SCRIPT_RESOURCE = "script.sh"


@dataclass(frozen=True, slots=True)
class LoginKeyfile:
    login: str
    has_pragma: bool
    readonly_condition: str | None
    keyfile: str


def _local_script_source() -> str:
    return resources.files(_SCRIPT_PACKAGE).joinpath(_SCRIPT_RESOURCE).read_text(encoding="utf-8")


def _local_script_sha256() -> str:
    return hashlib.sha256(_local_script_source().encode("utf-8")).hexdigest()


def _parse_keyfile_entry(raw: dict[str, Any]) -> LoginKeyfile:
    reason = raw.get("readonly_condition") or None
    return LoginKeyfile(
        login=str(raw["login"]),
        has_pragma=bool(raw["has_pragma"]),
        readonly_condition=reason if reason else None,
        # The shell script encodes "\n" as a literal two-char escape.
        keyfile=str(raw.get("keyfile", "")).replace("\\n", "\n"),
    )


class ScriptRunner:
    """High-level ops that delegate to the remote ``script.sh``."""

    def __init__(self, client: SshClient, *, remote_path: str = REMOTE_SCRIPT_PATH) -> None:
        self._client = client
        self._remote_path = remote_path

    async def ensure_uploaded(self, target: SshTarget) -> None:
        """Upload / refresh the script if the remote copy is missing or stale."""
        expected_sha = _local_script_sha256()
        probe = await self._client.exec(
            target,
            f"sh {shlex.quote(self._remote_path)} version 2>/dev/null || true",
        )
        if probe.stdout.strip():
            try:
                payload = json.loads(probe.stdout)
                if payload.get("sha256") == expected_sha:
                    return
            except json.JSONDecodeError:
                pass

        # Upload via a shell heredoc-free approach: stdin → cat → destination.
        await self._client.exec(
            target,
            (
                f"mkdir -p {shlex.quote(self._remote_path.rsplit('/', 1)[0])} "
                f"&& cat > {shlex.quote(self._remote_path)} "
                f"&& chmod 0700 {shlex.quote(self._remote_path)}"
            ),
            input=_local_script_source(),
        )

    async def get_ssh_keyfiles(self, target: SshTarget) -> list[LoginKeyfile]:
        """Return one :class:`LoginKeyfile` per login that has authorized_keys."""
        await self.ensure_uploaded(target)
        result = await self._client.exec(
            target, f"sh {shlex.quote(self._remote_path)} get_ssh_keyfiles"
        )
        if result.exit_code != 0:
            raise SshConnectFailed(
                f"script.sh get_ssh_keyfiles failed on {target.name}: {result.stderr.strip()}"
            )
        text = result.stdout.strip()
        if not text:
            return []
        try:
            payload = json.loads(text)
        except json.JSONDecodeError as exc:
            raise SshConnectFailed(f"script.sh returned non-JSON on {target.name}: {exc}") from exc
        if not isinstance(payload, list):
            raise SshConnectFailed(f"script.sh returned non-list on {target.name}")
        return [_parse_keyfile_entry(entry) for entry in payload]

    async def set_authorized_keyfile(self, target: SshTarget, *, login: str, content: str) -> None:
        """Replace ``login``'s ``authorized_keys`` with ``content``.

        Respects the readonly sentinel honored by the script itself; raises
        :class:`SshReadOnly` when the script refuses to overwrite.
        """
        await self.ensure_uploaded(target)
        command = f"sh {shlex.quote(self._remote_path)} set_authorized_keyfile {shlex.quote(login)}"
        result = await self._client.exec(target, command, input=content)
        if result.exit_code == 0:
            return
        stderr = result.stderr.strip() or result.stdout.strip()
        if "readonly" in stderr.lower():
            raise SshReadOnly(f"host {target.name!r} refused write for {login!r}: {stderr}")
        raise SshConnectFailed(
            f"script.sh set_authorized_keyfile failed on {target.name}: {stderr}"
        )

    async def version(self, target: SshTarget) -> dict[str, str]:
        """Return the remote script's ``{version, sha256}`` self-report."""
        result = await self._client.exec(target, f"sh {shlex.quote(self._remote_path)} version")
        if result.exit_code != 0:
            raise SshConnectFailed(
                f"script.sh version failed on {target.name}: {result.stderr.strip()}"
            )
        try:
            payload = json.loads(result.stdout)
        except json.JSONDecodeError as exc:
            raise SshConnectFailed(f"script.sh version non-JSON on {target.name}: {exc}") from exc
        return {str(k): str(v) for k, v in payload.items()}

File: phoenix/lib/ssm/ssh/script_runner.ex
defmodule Ssm.Ssh.ScriptRunner do
  @moduledoc """
  Drive `script.sh` on the remote host for every authorized_keys op — port of
  ssh/script_runner.py. The script itself (priv/ssh/script.sh) is carried
  over from the python stack VERBATIM; it handles home-directory lookup via
  getent (BSD/TrueNAS/pfSense), the readonly probe + vendor fingerprints, and
  has_pragma detection.

  All writes to authorized_keys MUST go through `set_authorized_keyfile/4` —
  never `write_file` directly — so the script can back up handwritten files
  and enforce the readonly flag.
  """

  alias Ssm.Ssh.{Result, Shell, Target}

  @remote_path ".ssm/script.sh"

  defmodule LoginKeyfile do
    @moduledoc "One login's authorized_keys as reported by script.sh."
    @enforce_keys [:login, :has_pragma, :keyfile]
    defstruct [:login, :has_pragma, :readonly_condition, :keyfile]

    @type t :: %__MODULE__{
            login: String.t(),
            has_pragma: boolean(),
            readonly_condition: String.t() | nil,
            keyfile: String.t()
          }
  end

  @doc "Upload / refresh the script if the remote copy is missing or stale."
  @spec ensure_uploaded(module(), Target.t()) :: :ok | {:error, term()}
  def ensure_uploaded(client \\ Ssm.Ssh, target) do
    probe_cmd = "sh #{Shell.quote(@remote_path)} version 2>/dev/null || true"

    with {:ok, %Result{} = probe} <- client.exec(target, probe_cmd, []) do
      if remote_current?(probe.stdout) do
        :ok
      else
        upload(client, target)
      end
    end
  end

  defp remote_current?(stdout) do
    case Jason.decode(String.trim(stdout)) do
      {:ok, %{"sha256" => sha}} -> sha == script_sha256()
      _ -> false
    end
  end

  defp upload(client, target) do
    dir = @remote_path |> Path.dirname() |> Shell.quote()

    command =
      "mkdir -p #{dir} && cat > #{Shell.quote(@remote_path)} && chmod 0700 #{Shell.quote(@remote_path)}"

    case client.exec(target, command, input: script_source()) do
      {:ok, %Result{exit_code: 0}} ->
        :ok

      {:ok, %Result{} = result} ->
        {:error,
         {:ssh_connect_failed,
          "script upload failed on #{target.name}: #{String.trim(result.stderr)}"}}

      {:error, _} = error ->
        error
    end
  end

  @doc "One `LoginKeyfile` per login that has an authorized_keys file."
  @spec get_ssh_keyfiles(module(), Target.t()) :: {:ok, [LoginKeyfile.t()]} | {:error, term()}
  def get_ssh_keyfiles(client \\ Ssm.Ssh, target) do
    with :ok <- ensure_uploaded(client, target),
         {:ok, %Result{} = result} <-
           client.exec(target, "sh #{Shell.quote(@remote_path)} get_ssh_keyfiles", []) do
      cond do
        result.exit_code != 0 ->
          {:error,
           {:ssh_connect_failed,
            "script.sh get_ssh_keyfiles failed on #{target.name}: #{String.trim(result.stderr)}"}}

        String.trim(result.stdout) == "" ->
          {:ok, []}

        true ->
          parse_keyfiles(result.stdout, target)
      end
    end
  end

  defp parse_keyfiles(stdout, target) do
    case Jason.decode(String.trim(stdout)) do
      {:ok, entries} when is_list(entries) ->
        {:ok, Enum.map(entries, &parse_entry/1)}

      {:ok, _other} ->
        {:error, {:ssh_connect_failed, "script.sh returned non-list on #{target.name}"}}

      {:error, error} ->
        {:error,
         {:ssh_connect_failed, "script.sh returned non-JSON on #{target.name}: #{inspect(error)}"}}
    end
  end

  defp parse_entry(raw) do
    readonly =
      case raw["readonly_condition"] do
        value when value in [nil, ""] -> nil
        value -> value
      end

    %LoginKeyfile{
      login: to_string(raw["login"]),
      has_pragma: !!raw["has_pragma"],
      readonly_condition: readonly,
      # The shell script encodes "\n" as a literal two-char escape.
      keyfile: raw |> Map.get("keyfile", "") |> to_string() |> String.replace("\\n", "\n")
    }
  end

  @doc """
  Replace `login`'s authorized_keys with `content`. The script enforces the
  readonly sentinel itself; its refusal surfaces as `{:error, {:ssh_readonly,
  reason}}`.
  """
  @spec set_authorized_keyfile(module(), Target.t(), String.t(), String.t()) ::
          :ok | {:error, term()}
  def set_authorized_keyfile(client \\ Ssm.Ssh, target, login, content) do
    command = "sh #{Shell.quote(@remote_path)} set_authorized_keyfile #{Shell.quote(login)}"

    with :ok <- ensure_uploaded(client, target),
         {:ok, %Result{} = result} <- client.exec(target, command, input: content) do
      cond do
        result.exit_code == 0 ->
          :ok

        readonly_refusal?(result) ->
          {:error,
           {:ssh_readonly,
            "host #{inspect(target.name)} refused write for #{inspect(login)}: " <>
              refusal_reason(result)}}

        true ->
          {:error,
           {:ssh_connect_failed,
            "script.sh set_authorized_keyfile failed on #{target.name}: " <>
              refusal_reason(result)}}
      end
    end
  end

  defp refusal_reason(%Result{} = result) do
    case String.trim(result.stderr) do
      "" -> String.trim(result.stdout)
      stderr -> stderr
    end
  end

  defp readonly_refusal?(%Result{} = result) do
    result |> refusal_reason() |> String.downcase() |> String.contains?("readonly")
  end

  @doc "The remote script's `{version, sha256}` self-report."
  @spec version(module(), Target.t()) :: {:ok, map()} | {:error, term()}
  def version(client \\ Ssm.Ssh, target) do
    with {:ok, %Result{} = result} <-
           client.exec(target, "sh #{Shell.quote(@remote_path)} version", []) do
      if result.exit_code == 0 do
        case Jason.decode(String.trim(result.stdout)) do
          {:ok, payload} when is_map(payload) ->
            {:ok, payload}

          _ ->
            {:error, {:ssh_connect_failed, "script.sh version non-JSON on #{target.name}"}}
        end
      else
        {:error,
         {:ssh_connect_failed,
          "script.sh version failed on #{target.name}: #{String.trim(result.stderr)}"}}
      end
    end
  end

  @doc "Local script source (priv/ssh/script.sh, verbatim from the python stack)."
  @spec script_source() :: String.t()
  def script_source do
    File.read!(Application.app_dir(:ssm, ["priv", "ssh", "script.sh"]))
  end

  @doc "SHA-256 of the local script, hex-encoded."
  @spec script_sha256() :: String.t()
  def script_sha256 do
    :crypto.hash(:sha256, script_source()) |> Base.encode16(case: :lower)
  end
end
